src/inspector_agent.cc

```diff
@@ -0,0 +1,515 @@
+    uv_async_send(&agent->data_written_);
+  } else if (read < 0) {
+    if (agent->client_socket_ == socket) {
+      agent->client_socket_ = nullptr;
+    }
+    DisconnectAndDisposeIO(socket);
+  } else {
+    // EOF
+    if (agent->client_socket_ == socket) {
+      agent->client_socket_ = nullptr;
+      agent->platform_->CallOnForegroundThread(agent->parent_env_->isolate(),
+          new SetConnectedTask(agent, false));
+      uv_async_send(&agent->data_written_);
+    }
+  }
+  uv_cond_broadcast(&agent->pause_cond_);
+}
+
+// static
+void Agent::WriteCbIO(uv_async_t* async) {
+  auto req = reinterpret_cast<AsyncWriteRequest*>(async->data);
+  req->perform();
+  delete req;
+  uv_close(reinterpret_cast<uv_handle_t*>(async), DisposeAsyncCb);
+}
+
+void Agent::WorkerRunIO() {
+  int err;
+  sockaddr_in addr;
+  uv_tcp_t server;
+  uv_tcp_init(&child_loop_, &server);
+  uv_ip4_addr("0.0.0.0", port_, &addr);
+  server.data = this;
+  err = uv_tcp_bind(&server, (const struct sockaddr*)&addr, 0);
+  if (err == 0) {
+    err = uv_listen(reinterpret_cast<uv_stream_t*>(&server), 0,
+                    OnSocketConnectionIO);
+  }
+  if (err == 0) {
+    PrintDebuggerReadyMessage(port_);
+  } else {
+    fprintf(stderr, "Unable to open devtools socket: %s\n", uv_strerror(err));
+    assert(false);
+  }
+  if (!wait_) {
+    uv_sem_post(&start_sem_);
+  }
+  uv_run(&child_loop_, UV_RUN_DEFAULT);
+  uv_close(reinterpret_cast<uv_handle_t*>(&server), nullptr);
+  uv_run(&child_loop_, UV_RUN_NOWAIT);
+}
+
+void Agent::OnInspectorConnectionIO(inspector_socket_t* socket) {
+  if (client_socket_) {
+    return;
+  }
+  client_socket_ = socket;
+  inspector_read_start(socket, OnBufferAlloc, Agent::OnRemoteDataIO);
+  platform_->CallOnForegroundThread(parent_env_->isolate(),
+      new SetConnectedTask(this, true));
+}
+
+void Agent::PostMessages() {
+  if (dispatching_messages_)
+    return;
+  dispatching_messages_ = true;
+  std::vector<blink::protocol::String16> messages;
+  uv_mutex_lock(&queue_lock_);
+  messages.swap(message_queue_);
+  uv_mutex_unlock(&queue_lock_);
+
+  for (auto const& message : messages)
+    inspector_->dispatchMessageFromFrontend(message);
+  uv_async_send(&data_written_);
+  dispatching_messages_ = false;
+}
+
+void Agent::SetConnected(bool connected) {
+  if (connected_ == connected)
+    return;
+
+  connected_ = connected;
+  if (connected) {
+    fprintf(stderr, "Debugger attached.\n");
+    inspector_->connectFrontend(new ChannelImpl(this));
+  } else {
+    PrintDebuggerReadyMessage(port_);
+    inspector_->quitMessageLoopOnPause();
+    inspector_->disconnectFrontend();
+  }
+}
+
+void Agent::Write(const String16& message) {
+  uv_async_t* async = reinterpret_cast<uv_async_t*>(malloc(sizeof(uv_async_t)));
+  ASSERT_NE(async, nullptr);
+  uv_async_init(&child_loop_, async, Agent::WriteCbIO);
+  async->data = new AsyncWriteRequest(this, message);
+  ASSERT_EQ(0, uv_async_send(async));
+}
```

Comment:
Does this function execute on a different thread than the one that calls `uv_run(&child_loop_, UV_RUN_DEFAULT)` runs?  If it does, the logic is wrong: it's not safe to call `uv_async_init()` on a different thread, you need to create it upfront on the uv_run thread.  Keep in mind that `uv_async_send()` calls can be coalesced into a single callback event.